### PR TITLE
Add check for zero to xrandom

### LINKED ISSUE #3039: Should xrandom be defensive against ZeroDivisionError?
Hi,

At the moment if the parameter provided for the percent in the `xrandom` random external trigger is zero, there will be a `ZeroDivisionError`.

```bash
2019-03-24T18:10:24+13:00 DEBUG - END TASK PROCESSING (took 0.002267599105834961 seconds)
2019-03-24T18:10:25+13:00 DEBUG - [xtrigger-func cmd] cylc-function-run xrandom '[]' '{"percent": 0, "secs": 2, "_": "1"}' /home/kinow/Development/python/workspace/cylc/etc/examples/xtrigger/xrandom
	[xtrigger-func ret_code] 1
	[xtrigger-func err]
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc/bin/cylc-function-run", line 37, in <module>
	    run_function(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 85, in run_function
	    res = func(*func_args, **func_kwargs)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/xtriggers/xrandom.py", line 44, in xrandom
	    satisfied = (1 == randint(1, 100 / int(percent)))
	ZeroDivisionError: division by zero
2019-03-24T18:10:25+13:00 ERROR - the JSON object must be str, bytes or bytearray, not NoneType
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 257, in start
	    self.run()
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 1534, in run
	    self.proc_pool.process()
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 174, in process
	    self._proc_exit(proc, "", ctx, callback, callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 165, in _proc_exit
	    self._run_command_exit(ctx, callback, callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 360, in _run_command_exit
	    callback(ctx, *callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/xtrigger_mgr.py", line 263, in callback
	    satisfied, results = json.loads(ctx.out)
	  File "/home/kinow/Development/python/anaconda3/lib/python3.7/json/__init__.py", line 341, in loads
	    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
	TypeError: the JSON object must be str, bytes or bytearray, not NoneType
2019-03-24T18:10:25+13:00 ERROR - error caught: cleaning up before exit
2019-03-24T18:10:25+13:00 INFO - Suite shutting down - ERROR: the JSON object must be str, bytes or bytearray, not NoneType
2019-03-24T18:10:25+13:00 ERROR - [Errno 3] No such process
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 257, in start
	    self.run()
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 1534, in run
	    self.proc_pool.process()
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 174, in process
	    self._proc_exit(proc, "", ctx, callback, callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 165, in _proc_exit
	    self._run_command_exit(ctx, callback, callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 360, in _run_command_exit
	    callback(ctx, *callback_args)
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/xtrigger_mgr.py", line 263, in callback
	    satisfied, results = json.loads(ctx.out)
	  File "/home/kinow/Development/python/anaconda3/lib/python3.7/json/__init__.py", line 341, in loads
	    raise TypeError(f'the JSON object must be str, bytes or bytearray, '
	TypeError: the JSON object must be str, bytes or bytearray, not NoneType
	
	During handling of the above exception, another exception occurred:
	
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 283, in start
	    self.shutdown('ERROR: ' + str(exc))
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/scheduler.py", line 1727, in shutdown
	    self.proc_pool.terminate()
	  File "/home/kinow/Development/python/workspace/cylc/lib/cylc/subprocpool.py", line 263, in terminate
	    os.killpg(proc.pid, SIGKILL)
	ProcessLookupError: [Errno 3] No such process
2019-03-24T18:10:25+13:00 INFO - DONE
```

Simply doing 

```diff
diff --git a/lib/cylc/xtriggers/xrandom.py b/lib/cylc/xtriggers/xrandom.py
index 43f4ca4cb..0a36c3b86 100644
--- a/lib/cylc/xtriggers/xrandom.py
+++ b/lib/cylc/xtriggers/xrandom.py
@@ -41,7 +41,7 @@ def xrandom(percent, secs=0, _=None, debug=False):
     """
     sleep(float(secs))
     results = {}
-    satisfied = (1 == randint(1, 100 / int(percent)))
+    satisfied = int(percent) != 0 and (1 == randint(1, 100 / int(percent)))
     if satisfied:
         results = {
             'COLOR': COLORS[randint(0, len(COLORS) - 1)],
```

Would return `False` when zero, however, the trigger would simply never succeed. So the following could be useful to prevent the `ZeroDivisionError`? If for whatever reason the percent is zero, then just consider it as done.

```diff
diff --git a/lib/cylc/xtriggers/xrandom.py b/lib/cylc/xtriggers/xrandom.py
index 43f4ca4cb..4793411d2 100644
--- a/lib/cylc/xtriggers/xrandom.py
+++ b/lib/cylc/xtriggers/xrandom.py
@@ -41,7 +41,7 @@ def xrandom(percent, secs=0, _=None, debug=False):
     """
     sleep(float(secs))
     results = {}
-    satisfied = (1 == randint(1, 100 / int(percent)))
+    satisfied = int(percent) == 0 or (1 == randint(1, 100 / int(percent)))
     if satisfied:
         results = {
             'COLOR': COLORS[randint(0, len(COLORS) - 1)],
```

### PR DESCRIPTION
**Pull request by kinow**
Fix #3039 

Prevents `ZeroDivisionError` in `xrandom`. Adds docstrings and doctests (thought a unittest for it wouldn't be really necessary as it's an example only? We were running doctests anyway for coverage...)

Cheers
Bruno

### CODE REVIEW

**Comment by kinow** [comment]
Output of `python -m doctest -v lib/cylc/xtriggers/xrandom.py`:

```bash
Trying:
    xrandom(0, 0)
Expecting:
    (False, {})
ok
Trying:
    import sys
Expecting nothing
ok
Trying:
    mocked_randint = lambda x, y: 10
Expecting nothing
ok
Trying:
    sys.modules[__name__].randint = mocked_randint
Expecting nothing
ok
Trying:
    xrandom(20, 0)
Expecting:
    (False, {})
ok
Trying:
    import sys
Expecting nothing
ok
Trying:
    mocked_randint = lambda x, y: 1
Expecting nothing
ok
Trying:
    sys.modules[__name__].randint = mocked_randint
Expecting nothing
ok
Trying:
    xrandom(1, 0)
Expecting:
    (True, {'COLOR': 'orange', 'SIZE': 'small'})
ok
1 items had no tests:
    xrandom
1 items passed all tests:
   9 tests in xrandom.xrandom
9 tests in 2 items.
9 passed and 0 failed.
Test passed.
```

**Comment by kinow** [comment]
I think it's not necessary to backport to 7.8.x, but happy to prepare another PR for the other branch if others prefer to include this fix there too.

**Comment by hjoliver** [comment]
@kinow - as a bug fix for existing code, this probably should be back-ported if you don't mind.

**Comment by hjoliver** [comment]
Codacy:
> Standard pseudo-random generators are not suitable for security/cryptographic purposes.

I wonder why Codacy sees this as a new issue? (same use of `random()` before and after this change).

**Comment by kinow** [comment]
>@kinow - as a bug fix for existing code, this probably should be back-ported if you don't mind.

Sure thing, pull request coming!

>I wonder why Codacy sees this as a new issue? (same use of random() before and after this change).

I suspect they are simply reporting on the old issue, as if it were a new one? We could add a #noqa marker and see if it goes away, or move the new `if` condition to before the existing one.

**Comment by hjoliver** [comment]
> We could add a #noqa marker ...

Can you try that?  (or should it be #nosec in this case?).  It's clearly not a security issue in this usage!

**Comment by kinow** [comment]
Roger. Before `bandit` reports:

```
kinow@kinow-VirtualBox:~/Development/python/workspace/cylc$ bandit lib/cylc/xtriggers/xrandom.py
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	running on Python 2.7.15
Run started:2019-03-25 00:12:48.244873

Test results:
>> Issue: [B311:blacklist] Standard pseudo-random generators are not suitable for security/cryptographic purposes.
   Severity: Low   Confidence: High
   Location: lib/cylc/xtriggers/xrandom.py:77
   More Info: https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b311-random
76	    results = {}
77	    satisfied = int(percent) != 0 and (1 == randint(1, 100 / int(percent)))
78	    if satisfied:

--------------------------------------------------
>> Issue: [B311:blacklist] Standard pseudo-random generators are not suitable for security/cryptographic purposes.
   Severity: Low   Confidence: High
   Location: lib/cylc/xtriggers/xrandom.py:80
   More Info: https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b311-random
79	        results = {
80	            'COLOR': COLORS[randint(0, len(COLORS) - 1)],
81	            'SIZE': SIZES[randint(0, len(SIZES) - 1)]

--------------------------------------------------
>> Issue: [B311:blacklist] Standard pseudo-random generators are not suitable for security/cryptographic purposes.
   Severity: Low   Confidence: High
   Location: lib/cylc/xtriggers/xrandom.py:81
   More Info: https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b311-random
80	            'COLOR': COLORS[randint(0, len(COLORS) - 1)],
81	            'SIZE': SIZES[randint(0, len(SIZES) - 1)]
82	        }

--------------------------------------------------

Code scanned:
	Total lines of code: 53
	Total lines skipped (#nosec): 0

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 3
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 3
Files skipped (0):
```

With (pushing in a commit in a bit):

```diff
diff --git a/lib/cylc/xtriggers/xrandom.py b/lib/cylc/xtriggers/xrandom.py
index ea0c48fa9..1b1b1e995 100644
--- a/lib/cylc/xtriggers/xrandom.py
+++ b/lib/cylc/xtriggers/xrandom.py
@@ -74,10 +74,11 @@ def xrandom(percent, secs=0, _=None, debug=False):
     """
     sleep(float(secs))
     results = {}
-    satisfied = int(percent) != 0 and (1 == randint(1, 100 / int(percent)))
+    satisfied = int(percent) != 0 and \
+                (1 == randint(1, 100 / int(percent)))  # nosec
     if satisfied:
         results = {
-            'COLOR': COLORS[randint(0, len(COLORS) - 1)],
-            'SIZE': SIZES[randint(0, len(SIZES) - 1)]
+            'COLOR': COLORS[randint(0, len(COLORS) - 1)],  # nosec
+            'SIZE': SIZES[randint(0, len(SIZES) - 1)]  # nosec
         }
     return (satisfied, results)
```

Results in:

```
[main]	INFO	profile include tests: None
[main]	INFO	profile exclude tests: None
[main]	INFO	cli include tests: None
[main]	INFO	cli exclude tests: None
[main]	INFO	running on Python 2.7.15
Run started:2019-03-25 00:14:16.326919

Test results:
	No issues identified.

Code scanned:
	Total lines of code: 54
	Total lines skipped (#nosec): 5

Run metrics:
	Total issues (by severity):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
	Total issues (by confidence):
		Undefined: 0
		Low: 0
		Medium: 0
		High: 0
Files skipped (0):
```

Let's wait and see what Codacy says after the commit is pushed.

**Comment by kinow** [comment]
Oh, and you were right, nosec, not noqa :grin:

**Comment by kinow** [comment]
@hjoliver looks it still complains about the added `if` statement, but not about the security issue with random:

>Up to standards. A positive pull request.

If that looks OK I will cherry-pick in the `7.8.x` branch and prepare a pull request.

**Comment by hjoliver** [comment]
That's great, plz go ahead with the back-port. Thanks!

**Comment by hjoliver** [comment]
@dwsutherland - you've worked with external trigger recently, so can you provide 2nd review for this?  (It'll only take a minute - just a small divide-by-zero bug fix, not really something that requires knowing about xtriggers actually).

File: lib/cylc/xtriggers/xrandom.py [hunk 1]
```diff
@@ -38,13 +38,47 @@ def xrandom(percent, secs=0, _=None, debug=False):
...
+    Returns:
+        A tuple with the trigger flag (bool) which is set to True if the
+        trigger was evaluated successful, False otherwise, and a random
+        color and size (dict).
     """
     sleep(float(secs))
     results = {}
-    satisfied = (1 == randint(1, 100 / int(percent)))
+    satisfied = int(percent) != 0 and \
+        (1 == randint(1, 100 / int(percent)))  # nosec
```
Comment (dwsutherland) [inline lib/cylc/xtriggers/xrandom.py]: This breaks with `randint(1, 100/int(99))` because it  essentially you're giving it float `1< N < 1.1111...`..  it requires `a <= N <= b` with `a` and `b` both integers.

You could incorporate the percentage 0 by doing the following:
```
    satisfied =  randint(1, 100) <= int(percent)
```
(Or `<= int(percent) % 100`, to ensure value between 0-100)

Comment (kinow) [reply lib/cylc/xtriggers/xrandom.py]: :open_mouth: ! Nice catch @dwsutherland ! I believe you found a new bug, and an interesting one!

And one that does not affect `7.8.x`, because in Python 2  `100/int(99)` is 1.

@dwsutherland , what if we simply change it to `(1 == randint(1, 100 // int(percent)))`? To maintain the old behaviour.

Comment (kinow) [reply lib/cylc/xtriggers/xrandom.py]:
```
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 100/int(99)
1
>>>
```

And

```
>>> from random import randint
>>> randint(1, 100/int(99))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/random.py", line 221, in randint
    return self.randrange(a, b+1)
  File "/usr/lib/python3.6/random.py", line 194, in randrange
    raise ValueError("non-integer stop for randrange()")
ValueError: non-integer stop for randrange()
```

Possibly would send another pull request, just to distinguish between this one that was backported to 7.8.x, and the different change for 8.0a1. How does that sound?

Comment (hjoliver) [reply lib/cylc/xtriggers/xrandom.py]: Ouch!!! Good spotting.  In fact it breaks with most values of `percent`!  (My fault, in the original 😬 )

Comment (dwsutherland) [reply lib/cylc/xtriggers/xrandom.py]: Ahh.. I see you have to make the numerator a float in python2
```
>>> 100./int(99)
1.0101010101010102
```
The issue I have with;
`(1 == randint(1, 100 // int(percent)))`
is that you can't get a true success rate for 99 or even 90 .. because it gets truncated.

Comment (hjoliver) [reply lib/cylc/xtriggers/xrandom.py]: Yes, my implementation is deficient, even in Python 2 - although it looks slightly less insane there at least.  It worked properly for my typical test cases, e.g. 10% ... but not in general!

**Comment by hjoliver** [comment]
How about this (Python 3):

```
random.random() < int(percent)/100
```
e.g. true with 20% likelihood, if a random float between 0 and 1 is less than 0.2.

**Comment by dwsutherland** [comment]
> How about this (Python 3):
> 
> ```
> random.random() < int(percent)/100
> ```
> e.g. true with 20% likelihood, if a random float between 0 and 1 is less than 0.2.

I did have a suggestion above which requires no division:
```
satisfied =  randint(1, 100) <= int(percent)
```
or
```
satisfied =  randint(1, 100) <= int(percent) % 100
```
Or am I missing something XD

**Comment by dwsutherland** [comment]
It also incorporates the 0

**Comment by dwsutherland** [comment]
Actually I do like your solution @hjoliver .. because if you remove the `int()' then it would work for decimals `54.35%`!

**Comment by hjoliver** [comment]
I missed your suggestion, sorry ... but what's wrong with division?! (`percent/100` is safe)

**Comment by dwsutherland** [comment]
> I missed your suggestion, sorry ... but what's wrong with division?! (`percent/100` is safe)
Yes yours is safe too!

File: lib/cylc/xtriggers/xrandom.py (hunk 1 above)
Comment (kinow) [reply lib/cylc/xtriggers/xrandom.py]: Then should we use 

`random.random() < int(percent)/100`

?

Fixes the `ZeroDivisionError` and the issue with `float` & `int`.

**Comment by kinow** [comment]
Looks like `random.random() < int(percent)/100` is safe and fixes this issue and the new one found by @dwsutherland :tada: updating pull request.

**Comment by kinow** [comment]
Done, ran `pycodestyle` and `bandit` before too just to make sure Travis would be happy (except for the flaky tests, of course).

And I think the code looks much nicer too! Reading the line that checks the likelihood is even easier to understand. :clap: great to have more :eyes: reviewing the code.

**Comment by hjoliver** [comment]
Yes, shows both how difficult and how critical reviewing is!  (I guess didn't bother to read that bit properly because I wrote it myself some time ago and so assumed it was right, haha 😟 ).  Thanks again to @dwsutherland  for checking properly (I think he has a maths background 🎉 )

File: lib/cylc/xtriggers/xrandom.py [hunk 2]
```diff
@@ -38,13 +38,48 @@ def xrandom(percent, secs=0, _=None, debug=False):
...
+        debug (bool): flag to enable debug information.
+    Returns:
+        A tuple with the trigger flag (bool) which is set to True if the
+        trigger was evaluated successful, False otherwise, and a random
+        color and size (dict).
     """
     sleep(float(secs))
     results = {}
-    satisfied = (1 == randint(1, 100 / int(percent)))
+    satisfied = random() < int(percent) / 100  # nosec
```
Comment (dwsutherland) [inline lib/cylc/xtriggers/xrandom.py]:
```suggestion
    satisfied = random() < float(percent) / 100  # nosec
```
This would allow for decimal percentages i.e.  `33.33333%`   (works in py2 or py3)

Comment (dwsutherland) [reply lib/cylc/xtriggers/xrandom.py]: `float(percent) / 100` and `float(percent) / 100.0` both seem to work BTW

**Comment by kinow** [comment]
Both pull requests updated, should be ready for review if Travis-CI doesn't fail (which my indicate it wants another kick).

**Comment by hjoliver** [comment]
💥
